remove proper dihedrals from structure generation

fgrunewald opened the pull request.

pckroon commented on polyply/src/minimizer.py:
```diff
@@ -77,8 +77,15 @@ def compute_dih(params, coords):
     """
-    dih_angle = dih(coords[0], coords[1], coords[2], coords[3])
-    return (dih_angle - float(params[1]))**2.0
+    # only optimize imporper dihedrals
+    # we filter all non-imporpers out here because
+    # the itp file parser doesn't distinguish them
```
This is missing the "why". Also, don't improper type 4 also have multiplicity?

fgrunewald replied:
good point but it's hard to tell what it does. I can't think of a use case at the moment

fgrunewald commented:
@ricalessandri can you use this branch to build both AA P3HT in GROMOS and OPLS PEG and check the structure is OK.

pckroon commented on polyply/src/minimizer.py:
```diff
@@ -77,7 +77,12 @@ def compute_dih(params, coords):
+    # treat multiplicity correctly; imporpers
+    # are needed for keeping conjugated systems falt;
+    # proper dihedrals will get adjusted in the energy
+    # minization
```
This should go to the docstring, which is now very sparse/incorrect.

ricalessandri replied:
btw, note the multiple *imporerp(s)* and *falt*

ricalessandri commented:
> @ricalessandri can you use this branch to build both AA P3HT in GROMOS and OPLS PEG and check the structure is OK.

For PEG there's no difference between `master` and `fix/115` and it looks reasonable. P3HT instead looks much better in `fix/115` (the thiophene rings are especially borked with the current `master`).

fgrunewald commented:
> > @ricalessandri can you use this branch to build both AA P3HT in GROMOS and OPLS PEG and check the structure is OK.
> 
> For PEG there's no difference between `master` and `fix/115` and it looks like a very reasonable structure to me. P3HT instead looks much better in `fix/115` (the thiophene rings are especially borked with the current `master`).

That's great! So at least this actually fixes things. It also greatly improves all-atom CHARMM alkanes as far as I can tell.

fgrunewald commented:
I've edited a couple of things to improve the geometry optimization:

- implemented a tolerance allowed in angles, bonds, dihedrals, which is also tested in the test
- added a seed to the initial random structure in the test, so it doesn't depend on the starting coordinates
- now the success depends on the overall values instead of only convergence of the optimizer. That should be much better because the convergence is good even when the structure is garbage. 
- fixed the doc-string etc. of the improper dihedral stuff

pckroon commented on polyply/src/minimizer.py:
```diff
@@ -138,6 +141,9 @@ def optimize_geometry(block, coords, inter_types=[]):
     block:  :class:vermouth.molecule.Block
     coords: dict
         dictionary of coordinates in form atom_name:np.ndarray
+    tolarance: dict
```
s/tolarance/tolerance/

Also in the signature

pckroon commented on polyply/src/minimizer.py:
```diff
@@ -166,11 +172,19 @@ def target_function(positions):
         coords[node_key] = positions[idx]
 
-    return opt_results['success'], coords
+    max_fun = 0
+    for inter_type, tol in tolarance.items():
+        max_fun += tol**2. * len(block.interactions[inter_type])
```
As a user I would interpret the tolerance differently. Instead of aggregating it all on one heap (where a very good bond could compensate a very bad dihedral or vice versa), I would expect the tolerance to be a limit *for every single interaction*. i.e. make sure that all bonds are within tolerance, and that all angles are, etc. Any interaction (not type!) not within tolerance should trigger failure.

fgrunewald replied:
Yeah you're right probably not a very smart idea. I'll change this

pckroon reviewed (changes requested):
One typo, and an opinion for you to decide on.

ricalessandri commented:
PEO again gives the same. P3HT is somewhat less good (?) but it's hard to judge; here's some representative snapshots

P3HT (only thiophenes highlighted) before 40aa1aac5fdf83d754659cafc17088db8eaeb6c4:
<img width="300" alt="before-" src="https://user-images.githubusercontent.com/10633291/112071813-96d97900-8b3e-11eb-93df-7156e4149da9.png">

P3HT after 40aa1aac5fdf83d754659cafc17088db8eaeb6c4:
<img width="353" alt="after-last-changes" src="https://user-images.githubusercontent.com/10633291/112071921-c5efea80-8b3e-11eb-9fe6-7ba1de637fd8.png">

fgrunewald commented:
@ricalessandri can you run this with the setting back_fudge = 1? It seems to me like in the second case the distances between the centers of the residues is shorter. But from this VMD image you can't really tell. And by before that commit you mean at commit 7841ca4 or really just one commit prior to 40aa1aa?

fgrunewald commented:
One more addition on this matter. Apparently now it takes 2 iterations to optimize P3HT to the constraints specified. I suppose that the tighter optimization leads to in principle a better P3HT but also a smaller size. One defect of the previous algorithm was that the alkane chain had the tendency to be longer than one would expect probably leading to this increase in the volume. I will implement Peters comment on the tolerances and then test P3HT again. Worst case AA P3HT needs a different step_fudge factor, but at least for me the two versions both energy minimize without problems. So we'll see. But a very interesting turn of events.

pckroon commented:
RE: tolerances - Take a look at the `callback` option for `scipy.optimize.minimize`. That's a callable function that will get called once per minimization iteration, and the minimization will terminate when that function returns `True`. I'm not quite sure what it'll do with convergence and such (since the objective value and the tolerances are no longer directly related), so some experimentation would be required.

fgrunewald commented:
I decided against a callback because effectively that means the target function gets evaluated twice only for the purpose of terminating earlier or having it "more clean". In general I find that the success is more dependent on the initial coordinates. If they are not good enough you won't converge to the accuracy we require. So it seems better to terminate and do a complete restart.

pckroon commented on polyply/src/minimizer.py:
```diff
@@ -173,18 +173,25 @@ def target_function(positions):
+            atom_coords = [positions[atom_to_idx[name]]
+                           for name in atoms]
+            penalty = INTER_METHODS[inter_type](params, atom_coords)
+            if inter_type in ["bonds", "constraints"] and penalty > 10000*tolarance[inter_type]**2.:
+                return False, coords
+            elif penalty > 10000*tolarance[inter_type]**2.:
```
I don't like these hardcoded force constants. (And the hardcoded functional form is debatable, but I can live with that)
Instead, make a global dict with interaction_type -> force constant that is used both here and by the objective function.
Or even bake them into the INTER_METHODS, e.g. using functools.partial

fgrunewald replied:
So I moved the constants to a global dict WEIGHTS, which is now flexible. But this is something I typically don't want to expose to the user and so I'm more than happy with it being somewhat hard-coded. Depending on the FF one could think to use different weights, but that is very much in the future.

pckroon reviewed (changes requested):
Tiny nitpick left.

I'm tempted to optimize and tune the minimization into oblivion even though that's probably not productive or useful. Here's two ideas though:
1) Tune the BFGS parameters. In particular limit the number of function evaluations/iterations and use the callback to check the convergence.
2) Use the thresholds/tolerances to set up constraints, and use a constrained minimizer (SLSQP)
3) Do the minimization using internal coordinates, and use the tolerances as bounds

Anyway, like I said, probably not useful.

fgrunewald commented:
For the time being it is indeed a bit over the top. We just need good enough fragments. But I agree that moving to internal coordinates is probably a good idea for a number of reasons. One being you can actually deal with the dihedral angles properly or at least better. Having said that it would be very cool if a vermouth molecule had an internal coordinate binding, where you can get the internal coordinates from the interactions. This would also make rebuilding coordinates much simpler for example.

ricalessandri commented:
Yes, indeed, the ` step_fudge` does what expected. But indeed I also have no problems minimizing a single chain either way.

Btw, I tried the `back_fudge`  out of curiosity; if set equal to 1 it gives pretty good monomer reconstruction, although it leads to quite some monomer-monomer overlap:
<img width="400" alt="P3HT_n6_gromos-BACKFUDGE1" src="https://user-images.githubusercontent.com/10633291/113083063-27e4cb80-91a1-11eb-8855-f89be682f7f5.png">

all in all, the command with `back_fudge=1` needs more steps of minimization than without `back_fudge`.

ricalessandri commented on polyply/src/minimizer.py:
```diff
@@ -154,11 +172,26 @@ def target_function(positions):
+            atom_coords = [positions[atom_to_idx[name]]
+                           for name in atoms]
+            penalty = INTER_METHODS[inter_type](params, atom_coords)
+            if inter_type in ["bonds", "constraints"] and penalty > 10000*tolarance[inter_type]**2.:
```
as a minor thing, I think you still have the "tolarance" typo throughout `minimizer.py`

fgrunewald replied:
damn, I thought I had removed that. Good catch!
